File: test/integration/AccessorEquivalenceMain.go
package main

import (
	"fmt"
	"math"
	"os"

	angle "uavcan_dsdl_generated/uavcan/si/unit/angle"
	sample "uavcan_dsdl_generated/uavcan/si/sample/temperature"
	uavtime "uavcan_dsdl_generated/uavcan/time"
	file "uavcan_dsdl_generated/uavcan/file"
	node "uavcan_dsdl_generated/uavcan/node"
	scalar "uavcan_dsdl_generated/uavcan/primitive/scalar"
	temperature "uavcan_dsdl_generated/uavcan/si/unit/temperature"
)

// An accessor against the body it stands in for: a getter answers what deserialise puts in the
// field, on a full buffer and on a short one, and a setter writes what deserialise reads back.
// Values are compared as bits, so a NaN meets itself; an integer round trip is also exact.
var rng uint32 = 0x9E3779B9

func fill(buffer []byte) {
	for i := range buffer {
		rng ^= rng << 13
		rng ^= rng >> 17
		rng ^= rng << 5
		buffer[i] = byte(rng)
	}
}

func bitsOf(v any) uint64 {
	switch x := v.(type) {
	case uint8:
		return uint64(x)
	case uint16:
		return uint64(x)
	case int16:
		return uint64(uint16(x))
	case uint64:
		return x
	case float32:
		return uint64(math.Float32bits(x))
	case float64:
		return math.Float64bits(x)
	}
	panic("a kind the driver does not compare")
}

// field deserialises the buffer into a fresh object and answers the member and whether it succeeded.
func checkField[V any](size int, get func([]byte) V, set func([]byte, V) error, field func([]byte) (V, bool), exact bool) bool {
	wire := make([]byte, size)
	fill(wire)
	ok := true
	full, rc := field(wire)
	ok = ok && rc && bitsOf(get(wire)) == bitsOf(full)
	short, rc := field(wire[:size/2])
	ok = ok && rc && bitsOf(get(wire[:size/2])) == bitsOf(short)
	out := make([]byte, size)
	v := get(wire)
	ok = ok && set(out, v) == nil
	back, rc := field(out)
	ok = ok && rc && bitsOf(get(out)) == bitsOf(back)
	ok = ok && (!exact || bitsOf(v) == bitsOf(back))
	ok = ok && set(out[:0], v) != nil
	return ok
}

// An element of a fixed array, through the index the accessor takes; one past the capacity reads
// as zero and cannot be set.
func checkElement[V any](size int, capacity int, get func([]byte, int) V, set func([]byte, int, V) error, element func([]byte, int) (V, bool)) bool {
	ok := true
	for i := 0; i < capacity; i++ {
		wire := make([]byte, size)
		fill(wire)
		full, rc := element(wire, i)
		ok = ok && rc && bitsOf(get(wire, i)) == bitsOf(full)
		ok = ok && bitsOf(get(wire, capacity)) == 0
		out := make([]byte, size)
		v := get(wire, i)
		ok = ok && set(out, i, v) == nil
		back, rc := element(out, i)
		ok = ok && rc && bitsOf(get(out, i)) == bitsOf(back)
		ok = ok && set(out, capacity, v) != nil
	}
	return ok
}

var failures = 0

func report(name string, same bool) {
	verdict := "same"
	if !same {
		verdict = "DIFFER"
		failures++
	}
	fmt.Printf("%-40s %s\n", name, verdict)
}

func main() {
	report("uavcan.node.Version",
		checkField(2, node.VersionGetMajor, node.VersionSetMajor, func(b []byte) (uint8, bool) {
			var o node.Version
			_, err := o.Deserialize(b)
			return o.Major, err == nil
		}, true) &&
			checkField(2, node.VersionGetMinor, node.VersionSetMinor, func(b []byte) (uint8, bool) {
				var o node.Version
				_, err := o.Deserialize(b)
				return o.Minor, err == nil
			}, true))
	report("uavcan.primitive.scalar.Integer16",
		checkField(2, scalar.Integer16GetValue, scalar.Integer16SetValue, func(b []byte) (int16, bool) {
			var o scalar.Integer16
			_, err := o.Deserialize(b)
			return o.Value, err == nil
		}, true))
	report("uavcan.primitive.scalar.Natural64",
		checkField(8, scalar.Natural64GetValue, scalar.Natural64SetValue, func(b []byte) (uint64, bool) {
			var o scalar.Natural64
			_, err := o.Deserialize(b)
			return o.Value, err == nil
		}, true))
	report("uavcan.primitive.scalar.Real16",
		checkField(2, scalar.Real16GetValue, scalar.Real16SetValue, func(b []byte) (float32, bool) {
			var o scalar.Real16
			_, err := o.Deserialize(b)
			return o.Value, err == nil
		}, false))
	report("uavcan.primitive.scalar.Real64",
		checkField(8, scalar.Real64GetValue, scalar.Real64SetValue, func(b []byte) (float64, bool) {
			var o scalar.Real64
			_, err := o.Deserialize(b)
			return o.Value, err == nil
		}, false))
	report("uavcan.si.unit.temperature.Scalar",
		checkField(4, temperature.ScalarGetKelvin, temperature.ScalarSetKelvin, func(b []byte) (float32, bool) {
			var o temperature.Scalar
			_, err := o.Deserialize(b)
			return o.Kelvin, err == nil
		}, false))
	report("uavcan.file.Error",
		checkField(2, file.ErrorGetValue, file.ErrorSetValue, func(b []byte) (uint16, bool) {
			var o file.Error
			_, err := o.Deserialize(b)
			return o.Value, err == nil
		}, true))
	report("uavcan.si.unit.angle.Quaternion",
		checkElement(16, 4, angle.QuaternionGetWxyz, angle.QuaternionSetWxyz, func(b []byte, i int) (float32, bool) {
			var o angle.Quaternion
			_, err := o.Deserialize(b)
			return o.Wxyz[i], err == nil
		}))
	// A nested composite, through the buffer its getter answers: the nested type's own getter on it
	// agrees with deserialise on the full buffer and on one cut inside the nested field.
	{
		wire := make([]byte, 11)
		fill(wire)
		var obj sample.Scalar
		_, err := obj.Deserialize(wire)
		same := err == nil && uavtime.SynchronizedTimestampGetMicrosecond(sample.ScalarGetTimestamp(wire)) == obj.Timestamp.Microsecond
		var short sample.Scalar
		_, err = short.Deserialize(wire[:3])
		stamp := sample.ScalarGetTimestamp(wire[:3])
		same = same && err == nil && len(stamp) == 3 && uavtime.SynchronizedTimestampGetMicrosecond(stamp) == short.Timestamp.Microsecond
		same = same && checkField(11, sample.ScalarGetKelvin, sample.ScalarSetKelvin, func(b []byte) (float32, bool) {
			var o sample.Scalar
			_, err := o.Deserialize(b)
			return o.Kelvin, err == nil
		}, false)
		report("uavcan.si.sample.temperature.Scalar", same)
	}
	if failures != 0 {
		os.Exit(1)
	}
}

File: test/integration/AccessorEquivalenceMain.rs
// An accessor against the body it stands in for: a getter answers what deserialise puts in the
// field, on a full buffer and on a short one, and a setter writes what deserialise reads back.
// Values are compared as bits, so a NaN meets itself; an integer round trip is also exact.
use uavcan_dsdl_generated::uavcan::file::error_1_0::Error;
use uavcan_dsdl_generated::uavcan::node::version_1_0::Version;
use uavcan_dsdl_generated::uavcan::primitive::scalar::integer16_1_0::Integer16;
use uavcan_dsdl_generated::uavcan::primitive::scalar::natural64_1_0::Natural64;
use uavcan_dsdl_generated::uavcan::primitive::scalar::real16_1_0::Real16;
use uavcan_dsdl_generated::uavcan::primitive::scalar::real64_1_0::Real64;
use uavcan_dsdl_generated::uavcan::si::sample::temperature::scalar_1_0::Scalar as SampleTemperatureScalar;
use uavcan_dsdl_generated::uavcan::si::unit::angle::quaternion_1_0::Quaternion;
use uavcan_dsdl_generated::uavcan::time::synchronized_timestamp_1_0::SynchronizedTimestamp;
use uavcan_dsdl_generated::uavcan::si::unit::temperature::scalar_1_0::Scalar as UnitTemperatureScalar;

static mut RNG: u32 = 0x9E37_79B9;

fn fill(buffer: &mut [u8]) {
    for b in buffer.iter_mut() {
        // Single-threaded, and the state is read and written in one place.
        let mut s = unsafe { RNG };
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        unsafe { RNG = s };
        *b = s as u8;
    }
}

trait Bits {
    fn bits(&self) -> u64;
}
impl Bits for u8 {
    fn bits(&self) -> u64 {
        *self as u64
    }
}
impl Bits for u16 {
    fn bits(&self) -> u64 {
        *self as u64
    }
}
impl Bits for i16 {
    fn bits(&self) -> u64 {
        *self as u16 as u64
    }
}
impl Bits for u64 {
    fn bits(&self) -> u64 {
        *self
    }
}
impl Bits for f32 {
    fn bits(&self) -> u64 {
        self.to_bits() as u64
    }
}
impl Bits for f64 {
    fn bits(&self) -> u64 {
        self.to_bits()
    }
}

macro_rules! check_field {
    ($t:ty, $field:ident, $get:expr, $set:expr, $size:expr, $exact:expr) => {{
        let mut wire = vec![0u8; $size];
        fill(&mut wire);
        let mut ok = true;
        let (obj, _) = <$t>::from_bytes(&wire).unwrap();
        ok &= $get(&wire).bits() == obj.$field.bits();
        let half = $size / 2;
        let (short, _) = <$t>::from_bytes(&wire[..half]).unwrap();
        ok &= $get(&wire[..half]).bits() == short.$field.bits();
        let mut out = vec![0u8; $size];
        let v = $get(&wire);
        ok &= $set(&mut out, v).is_ok();
        let (back, _) = <$t>::from_bytes(&out).unwrap();
        ok &= $get(&out).bits() == back.$field.bits();
        if $exact {
            ok &= back.$field.bits() == v.bits();
        }
        ok &= $set(&mut out[..0], v).is_err();
        ok
    }};
}

// An element of a fixed array, through the index the accessor takes; one past the capacity reads
// as zero and cannot be set.
macro_rules! check_element {
    ($t:ty, $field:ident, $get:expr, $set:expr, $size:expr, $capacity:expr) => {{
        let mut ok = true;
        for i in 0..$capacity {
            let mut wire = vec![0u8; $size];
            fill(&mut wire);
            let (obj, _) = <$t>::from_bytes(&wire).unwrap();
            ok &= $get(&wire, i).bits() == obj.$field[i].bits();
            ok &= $get(&wire, $capacity).bits() == 0;
            let mut out = vec![0u8; $size];
            let v = $get(&wire, i);
            ok &= $set(&mut out, i, v).is_ok();
            let (back, _) = <$t>::from_bytes(&out).unwrap();
            ok &= $get(&out, i).bits() == back.$field[i].bits();
            ok &= $set(&mut out, $capacity, v).is_err();
        }
        ok
    }};
}

fn report(name: &str, same: bool) -> bool {
    println!("{:<40} {}", name, if same { "same" } else { "DIFFER" });
    same
}

fn main() {
    let mut ok = true;
    ok &= report(
        "uavcan.node.Version",
        check_field!(Version, major, Version::get_major, Version::set_major, 2, true)
            && check_field!(Version, minor, Version::get_minor, Version::set_minor, 2, true),
    );
    ok &= report(
        "uavcan.primitive.scalar.Integer16",
        check_field!(Integer16, value, Integer16::get_value, Integer16::set_value, 2, true),
    );
    ok &= report(
        "uavcan.primitive.scalar.Natural64",
        check_field!(Natural64, value, Natural64::get_value, Natural64::set_value, 8, true),
    );
    ok &= report(
        "uavcan.primitive.scalar.Real16",
        check_field!(Real16, value, Real16::get_value, Real16::set_value, 2, false),
    );
    ok &= report(
        "uavcan.primitive.scalar.Real64",
        check_field!(Real64, value, Real64::get_value, Real64::set_value, 8, false),
    );
    ok &= report(
        "uavcan.si.unit.temperature.Scalar",
        check_field!(UnitTemperatureScalar, kelvin, UnitTemperatureScalar::get_kelvin, UnitTemperatureScalar::set_kelvin, 4, false),
    );
    ok &= report(
        "uavcan.file.Error",
        check_field!(Error, value, Error::get_value, Error::set_value, 2, true),
    );
    ok &= report(
        "uavcan.si.unit.angle.Quaternion",
        check_element!(Quaternion, wxyz, Quaternion::get_wxyz, Quaternion::set_wxyz, 16, 4),
    );
    // A nested composite, through the buffer its getter answers: the nested type's own getter on it
    // agrees with deserialise on the full buffer and on one cut inside the nested field.
    {
        let mut wire = vec![0u8; 11];
        fill(&mut wire);
        let (obj, _) = SampleTemperatureScalar::from_bytes(&wire).unwrap();
        let stamp = SampleTemperatureScalar::get_timestamp(&wire);
        let mut same = SynchronizedTimestamp::get_microsecond(stamp) == obj.timestamp.microsecond;
        let (short, _) = SampleTemperatureScalar::from_bytes(&wire[..3]).unwrap();
        let stamp = SampleTemperatureScalar::get_timestamp(&wire[..3]);
        same &= stamp.len() == 3
            && SynchronizedTimestamp::get_microsecond(stamp) == short.timestamp.microsecond;
        same &= check_field!(SampleTemperatureScalar, kelvin, SampleTemperatureScalar::get_kelvin, SampleTemperatureScalar::set_kelvin, 11, false);
        ok &= report("uavcan.si.sample.temperature.Scalar", same);
    }
    std::process::exit(if ok { 0 } else { 1 });
}
